okhttp/src/main/java/okhttp3/Request.kt

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright (C) 2013 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3
+
+import okhttp3.internal.Util
+import okhttp3.internal.http.HttpMethod
+import java.net.URL
+
+/**
+ * An HTTP request. Instances of this class are immutable if their [body] is `null` or itself
+ * immutable.
+ */
+class Request internal constructor(
+  internal val url: HttpUrl,
```

Comment:
just curious: why does URL get constructor treatment but the other vals do not?

Reply:
Because this is the only one that is technically nullable in the builder property but required to not be by the time it gets here. It's checked (and thus smart casted safely) in the builder, but we could move that check to the constructor here instead if you'd rather keep it single-param